crates/biome_cli/src/execute/traverse.rs

```diff
@@ -204,79 +88,30 @@ pub(crate) fn traverse(
             },
         );
         // wait for the main thread to finish
-        handler.join().unwrap();
+        let diagnostics = handler.join().unwrap();
 
-        elapsed
+        (elapsed, diagnostics)
     });
 
     let errors = printer.errors();
     let warnings = printer.warnings();
     let changed = changed.load(Ordering::Relaxed);
     let unchanged = unchanged.load(Ordering::Relaxed);
-    let count = changed + unchanged;
     let skipped = skipped.load(Ordering::Relaxed);
-
-    if execution.should_report_to_terminal() {
-        console.log(markup! {
-            {SummaryResult {
-                changed,
-                unchanged,
-                duration,
-                errors,
-                warnings,
-                traversal: execution.traversal_mode()
-            }}
-        });
-    } else {
-        if let TraversalMode::Format { write, .. } = execution.traversal_mode() {
-            let mut summary = FormatterReportSummary::default();
-            if *write {
-                summary.set_files_written(count);
-            } else {
-                summary.set_files_compared(count);
-            }
-            report.set_formatter_summary(summary);
-        }
-
-        let to_print = report.as_serialized_reports()?;
-        console.log(markup! {
-            {to_print}
-        });
-        return Ok(());
-    }
-
-    if skipped > 0 {
-        if skipped == 1 {
-            console.log(markup! {
-                <Warn>"Skipped "{skipped}" file."</Warn>
-            });
-        } else {
-            console.log(markup! {
-                <Warn>"Skipped "{skipped}" files."</Warn>
-            });
-        }
-    }
-
-    let should_exit_on_warnings = warnings > 0 && cli_options.error_on_warnings;
-    // Processing emitted error diagnostics, exit with a non-zero code
-    if count.saturating_sub(skipped) == 0 && !cli_options.no_errors_on_unmatched {
-        Err(CliDiagnostic::no_files_processed())
-    } else if errors > 0 || should_exit_on_warnings {
-        let category = execution.as_diagnostic_category();
-        if should_exit_on_warnings {
-            if execution.is_check_apply() {
-                Err(CliDiagnostic::apply_warnings(category))
-            } else {
-                Err(CliDiagnostic::check_warnings(category))
-            }
-        } else if execution.is_check_apply() {
-            Err(CliDiagnostic::apply_error(category))
-        } else {
-            Err(CliDiagnostic::check_error(category))
-        }
-    } else {
-        Ok(())
-    }
```

Comment:
Nice cleanup! I like that these concerns got nicely separated in the `Reporter` trait.

Reply:
Same there! Looks cleaner 🤩.